### PR TITLE
Add Dokka plugin to exclude @Internal code from documentation 2.0

### PR DESCRIPTION
**Pull request by tripolkaandrey**
This PR introduces a new module `dokka-extensions`. There is a [custom Dokka plugin](https://kotlin.github.io/dokka/1.6.10/developer_guide/introduction/#building-sample-plugin) to extract from documentation information about the `@Internal` code.

Dokka has several [extension points](https://kotlin.github.io/dokka/1.6.10/developer_guide/extension_points/) at different stages of generating documentation. This new plugin injects a “transformer”(a term native to Dokka) at the stage when all code from source sets was collected and translated to internal Dokka representation [`Documentable`](https://kotlin.github.io/dokka/1.6.10/developer_guide/data_model/#documentable-model). This “transformer” extends class which does all the “messy” job of different processing of modules, packages, types, etc. leaving to us only overwriting the criteria by which we should suppress(exclude) some `Documentable`’s. 

Dokka [automatically loads](https://kotlin.github.io/dokka/1.6.10/developer_guide/introduction/#:~:text=All%20instances%20are%20automatically%20loaded%20during%20Dokka%20setup%20using%20java.util.ServiceLoader) this plugin if it is present on the classpath. If your project already has Dokka, this plugin can be added via 
```build.gradle.kts
dependencies {
    dokkaPlugin("io.spine.tools:spine-dokka-extensions:${version}")
}
```

Last but not least, this repo is no longer only about `Javadoc` tools. Now it is `doc-tools` because more than one documentation tool is involved, so required changes were made.

### CODE REVIEW

File: dokka-extensions/src/main/kotlin/io/spine/tools/dokka/plugin/ExcludeInternalPlugin.kt [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+ * @see <a href="https://kotlin.github.io/dokka/1.6.10/developer_guide/introduction/">
+ *     Guide to Dokka Plugin development</a>
+ *
+ * @see <a href="https://kotlin.github.io/dokka/1.6.10/developer_guide/extension_points/#pre-merge-documentation-transform">
+ *     Pre-merge documentation transform</a>
+ */
+public class ExcludeInternalPlugin : DokkaPlugin() {
+    private val dokkaBase by lazy { plugin<DokkaBase>() }
+
+    public val excludeInternalTransformer: Extension<PreMergeDocumentableTransformer, *, *> by extending {
```
Comment (armiol) [inline dokka-extensions/src/main/kotlin/io/spine/tools/dokka/plugin/ExcludeInternalPlugin.kt:59]: Please document this item. It's a part of `public` API.

File: dokka-extensions/src/main/kotlin/io/spine/tools/dokka/plugin/ExcludeInternalTransformer.kt [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+    /**
+     * Tells whether the passed [Documentable] should be excluded from the documentation.
+     *
+     * Returns `true` if the passed piece is annotated with the [Internal] annotation. Otherwise,
+     * returns `false`.
+     */
+    override fun shouldBeSuppressed(d: Documentable): Boolean {
+        /**
+         * [WithExtraProperties] is a Dokka-interface that adds a map-field for extra properties.
```
Comment (armiol) [inline dokka-extensions/src/main/kotlin/io/spine/tools/dokka/plugin/ExcludeInternalTransformer.kt:53]: Usually, we don't have block comments for the in-line expressions, other than to comment suppressions.

There are some exceptions, but this is not the one.

If you want to comment on the implementation details, please move this text to the method-level docs. There is an `@implNote` tag in Javadocs for that. AFAIK, KDoc doesn't allow that. In any case, the explanation may be just the continuation of the documentation you already have in place for the method.

File: dokka-extensions/src/main/kotlin/io/spine/tools/dokka/plugin/ExcludeInternalTransformer.kt [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+    /**
+     * Extracts annotations from a [WithExtraProperties] object.
+     *
+     * Information about annotations is found in the map-field `extra`, added when implementing
+     * [WithExtraProperties].
+     */
+    private fun WithExtraProperties<*>.annotations(): List<Annotation> {
+        /**
+         * All values with the type of [Annotations], a container-class for all applied annotations,
```
Comment (armiol) [inline dokka-extensions/src/main/kotlin/io/spine/tools/dokka/plugin/ExcludeInternalTransformer.kt:72]: Same as above.

File: dokka-extensions/src/main/kotlin/io/spine/tools/dokka/plugin/ExcludeInternalTransformer.kt [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+     * Provides the method to check if an [Annotation] object represents [Internal] annotation.
+     */
+    private object InternalAnnotationCheck {
+        private val c = Internal::class.java
+
+        /**
+         * Checks if provided object represents [Internal] annotation.
+         *
+         * Compares the package and the class extracted from the object with the package
+         * and the class of the [Internal] annotation.
```
Comment (armiol) [inline dokka-extensions/src/main/kotlin/io/spine/tools/dokka/plugin/ExcludeInternalTransformer.kt:91]: Judging on this description, I have doubts on whether the whole thing would work for the packages which are annotated with `@Internal` in their `package-info.java`. Let's check that out of scope of this PR (please create a respective GitHub issue in this repository).

File: pom.xml [hunk 5]
```diff
@@ -53,6 +53,12 @@ all modules and does not describe the project structure per-subproject.
     <version>1.3.2</version>
     <scope>compile</scope>
   </dependency>
+  <dependency>
+    <groupId>org.jetbrains.dokka</groupId>
+    <artifactId>dokka-base</artifactId>
+    <version>1.5.0</version>
```
Comment (armiol) [inline pom.xml:59]: As discussed previously, we need to grab the latest updates from `config`. So that here we would have the latest version.

**Review by armiol — CHANGES_REQUESTED** [review]
@tripolkaandrey we are getting closer. Please see my comments.

File: pom.xml (hunk 5 above)
Comment (tripolkaandrey) [reply pom.xml:59]: Yes, I remember that. I have pulled the config and am in the process of cleaning unresolved references. I will generate the pom and the license report again later.

File: dokka-extensions/src/main/kotlin/io/spine/tools/dokka/plugin/ExcludeInternalTransformer.kt (hunk 4 above)
Comment (tripolkaandrey) [reply dokka-extensions/src/main/kotlin/io/spine/tools/dokka/plugin/ExcludeInternalTransformer.kt:91]: I don't see why that should not work. Dokka collects annotations applied to the piece of code(whether it is a package, a type, or a method). We then extract this information as `List<Annotation>` and go one by one through the list checking if the current annotation represents `@Internal`. The check algorithm is described in the comments you mentioned.

Comment (tripolkaandrey) [reply dokka-extensions/src/main/kotlin/io/spine/tools/dokka/plugin/ExcludeInternalTransformer.kt:91]: I was wrong. By analyzing the source code of Dokka, I found out that they indeed do not check packages with `shouldBeSupressed`, which we overrode a few lines above. They immediately go to processing types, methods, etc. If it was not the case, this check would work as it is concept-agnostic and just compares provided annotation with `@Internal`.

**Pull request closed without merging**